### ISSUE TITLE
TimerService.stopService should wait for a running call to finish.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6290]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)| @tomprince reported|
|-|-|
|Trac ID|trac#6290|
|Type|defect|
|Created|2013-02-01 05:12:16Z|
|Branch|https://github.com/twisted/twisted/tree/TimerService-wait-for-finish-6290|

When `TimerService.stopService` is called while a call is in progress,
it should return a deferred that fires after the call finishes.

Attachments:

* [TimerService.stopService.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3ca/3cac1948eabda6e7a458fc1875a1a25c217e82af/83760ec70206118a1657ba5e227874719e994929.patch) (3388 bytes) - added by tom.prince on 2013-02-01 05:13:04Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6290 6290
type__defect defect
reporter__tom_prince tom.prince
priority__normal normal
milestone__ 
branch__branches_TimerService_wait_for_finish_6290 branches/TimerService-wait-for-finish-6290
branch_author__tom_prince tom.prince
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1359695536000000 1359695536000000
changetime__1362190599000000 1362190599000000
version__None None
owner__tom_prince tom.prince
cc__rwall
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_1"></a>|@tomprince commented|
|-|-|

(In [37160]) Branching to TimerService-wait-for-finish-6290.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_2"></a>|@tomprince removed owner|
|-|-|

[buildbot run](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/TimerService-wait-for-finish-6290)

The failing tests on lucid32-py2.7maint are due to [#6314](https://github.com/twisted/twisted/issues/6314) and are unrelated.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_3"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

Reviewing...

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_4"></a>|@wallrj set owner to @tomprince<br>@wallrj set status to `new`|
|-|-|

Looks pretty good Tom.

 * Tests all pass.
 * Merges cleanly to trunk.
 * trial --coverage shows coverage for all the changes in TimerService.

A few points:

 1. Missing news file
 2. [branches/TimerService-wait-for-finish-6290/twisted/application/internet.py](branches/TimerService-wait-for-finish-6290/twisted/application/internet.py)
    1. TimerService.clock:
       1. Shouldn't this be private? It's only
          used for testing purposes. Although I notice that LoopingCall
          has a public clock attribute. So maybe...
       2. Document it with @ivar, like LoopingCall does.
       3. ...in which case, fix the class docstring formatting while you're there.
       4. Should clock be listed as volatile to prevent it being
          (un)pickled / (un)serialised?
          In another service Procmon.ProcessMonitor - for example - I
          think exarkun asked me to
          explicitly delete the reactor attribute in __get_state__ which
          is a similar situation...maybe.
    2. TimeService.stopService
       1. IMHO it looks strange to upcall to
          service.Service.stopService in a callback. What if in the
          future, the base method needs to do something synchronously?
          Would it be safer to call stopService and assign the result
          to a local variable which can be returned from the callback?
          Not sure.
       2. I think you should override the docstring in this method.
          Explain that it *always* returns a deferred and that it
          returns when the any in progress calls have finished.
 3. [branches/TimerService-wait-for-finish-6290/twisted/application/test/test_internet.py](branches/TimerService-wait-for-finish-6290/twisted/application/test/test_internet.py)
    1. test_failedCallLogsError
       1. Typo in docstring "returns a deferred the errbacks"
       2. Why do you need clock.advance here? test_stopServiceWaits
          doesn't, it relies on LoopingCall.start(now=True), so why
          not rely on that here too? In which case, maybe you can get rid
          of TimerService.clock (for now). See what you think.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

> trial --coverage shows coverage for all the changes in TimerService.

Here's a tip that should probably be documented somewhere - `trial --coverage ...` is way worse than `python-coverage run --branch --source ... trial ...`.  You definitely always want to use the latter.  In fact, we should almost certainly get rid of the former (and _maybe_ replace it with something based on python-coverage, or maybe not).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_6"></a>|@tomprince set owner to @wallrj|
|-|-|

> 1. Missing news file
Done
> TimerService.clock:
> 2.1.1. Shouldn't this be private? It's only used for testing purposes. Although I notice that LoopingCall has a public clock attribute. So maybe...
We want to promote people using twisted to write testable code. Having this public allows
people to test the interaction of a `TimerService` with other code they write. (Note: They shouldn't be testing that `TimerService` works, but they may have functionality that depends on the operating of a `TimerService`)
> 2.1.2 Document it with @ivar, like LoopingCall does.
Done.
> 2.1.3 ...in which case, fix the class docstring formatting while you're there.
Done.
> 2.1.4 Should clock be listed as volatile to prevent it being (un)pickled / (un)serialised? In another service Procmon.ProcessMonitor - for example - I think exarkun asked me to explicitly delete the reactor attribute in get_state which is a similar situation...maybe. 
If the global reactor is being used, this is probably `None`, so I don't think that is necessary.
> 2.2 TimeService.stopService
> 2.2.1 IMHO it looks strange to upcall to service.Service.stopService in a callback. What if in the future, the base method needs to do something synchronously? Would it be safer to call stopService and assign the result to a local variable which can be returned from the callback? Not sure.
I want to upcall *after* this class has stopped shutting down.
> 2.2.2 I think you should override the docstring in this method. Explain that it *always* returns a deferred and that it returns when the any in progress calls have finished. 
Done.
> 3.1.        test_failedCallLogsError
> 3.1.1 Typo in docstring "returns a deferred the errbacks"
Fixed
> 3.1.2 Why do you need clock.advance here? test_stopServiceWaits doesn't, it relies on LoopingCall.start(now=True), so why not rely on that here too?
Done 
> In which case, maybe you can get rid of TimerService.clock (for now). See what you think.
I think it is a useful attribute, so I added some explicit tests for it. One thing about using `Clock` instead of the global reactor is that you don't need to be careful to cleanup after the tests (cancelling `callLater`s for example.)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_7"></a>|@wallrj set owner to @tomprince|
|-|-|

Ok. That all sounds fine, but...

 1. [Build results look bad](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/TimerService-wait-for-finish-6290)
 2. I noticed a few typos in the new docstrings
    1. periodcally
    2. secons
    3. triggered when the when
 3. Epylinks - I still don't properly understand the rules for these,
    but...
    1. L{None}  shouldn't that be C{None} cos there's nothing to
       actually link to.
    2. Same with L{float}, L{callable}
    3. L{start} should be L{TimerService.startService}

+1 from me if you make the build results good and if you fix those typos
and links.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_8"></a>|@wallrj commented|
|-|-|

...and "it's" > "its" in the news file.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_9"></a>|@tomprince set status to `closed`|
|-|-|

(In [37359]) Merge TimerService-wait-for-finish-6290: Make TimerService.stopService wait for a running call to finish.

Author: tom.prince
Reviewers: rwall
Fixes: [#6290](https://github.com/twisted/twisted/issues/6290)

There was no public way to wait for the running call
to finish which meant that reactor wouldn't wait for
it to complete, and that test couldn't clean up.
